File: homework3/Satyam_Mishra/Q11_VacationDestinations.java
// DS/Algo used - Dijkstra's Algorithm

// Time Complexity = O(ElogV)
// Space Complexity = O(V + E)

package homework3.Satyam_Mishra;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.PriorityQueue;
import java.util.Set;

class Triplet {
    String city1;
    String city2;
    double time;

    public Triplet(String city1, String city2, double time) {
        this.city1 = city1;
        this.city2 = city2;
        this.time = time;
    }
}

class Pair {
    String city;
    double time;

    public Pair(String city, double time) {
        this.city = city;
        this.time = time;
    }
}

public class Q11_VacationDestinations {
    private static List<String> bfs(String origin, int k, Map<String, List<Pair>> adjMap) {
        List<String> res = new ArrayList<>();
        Set<String> visited = new HashSet<>();

        PriorityQueue<Pair> pq = new PriorityQueue<>((a, b) -> Double.compare(a.time, b.time));
        pq.add(new Pair(origin, 0));

        while (!pq.isEmpty()) {
            Pair top = pq.poll();
            String city = top.city;
            double time = top.time;

            if(visited.contains(city)) continue;

            visited.add(city);
            if(!city.equals(origin)) res.add(city);

            for(Pair nei : adjMap.get(city)) {
                double newTime = time + nei.time;

                if(!city.equals(origin)) {
                    newTime += 1;
                }

                if(newTime > k) continue;

                pq.add(new Pair(nei.city, newTime));
            }
        }

        return res;
    }
    public static void main(String[] args) {
        List<Triplet> edges = new ArrayList<>();
        edges.add(new Triplet("Boston", "New York", 4));
        edges.add(new Triplet("New York", "Philadelphia", 2));
        edges.add(new Triplet("Boston", "Newport", 1.5));
        edges.add(new Triplet("Washington, D.C.", "Harper's Ferry", 1));
        edges.add(new Triplet("Boston", "Portland", 2.5));
        edges.add(new Triplet("Philadelphia", "Washington, D.C.", 2.5));

        String origin = "New York";
        int k = 5;

        Map<String, List<Pair>> adjMap = new HashMap<>();

        for (Triplet edge : edges) {
            adjMap.putIfAbsent(edge.city1, new ArrayList<>());
            adjMap.putIfAbsent(edge.city2, new ArrayList<>());

            adjMap.get(edge.city1).add(new Pair(edge.city2, edge.time));
            adjMap.get(edge.city2).add(new Pair(edge.city1, edge.time));
        }

        System.out.println(bfs(origin, k, adjMap));
    }
}

// time taken 22m 40s

File: homework3/Madhur/Q11_VacationDestinations.cpp
/*
Data Structure Used:
Graph (Adjacency List), Priority Queue (Min Heap), Hashmap

Algorithm:
Dijkstra’s Algorithm

Approach:
Use Dijkstra’s algorithm to compute the shortest travel time from the origin to all cities while adding stopover costs.

Time Complexity:
- O((V + E) log V)

Space Complexity:
- O(V + E)

Time Taken:
40 minutes
*/

#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    int vacationDestinations(
        vector<tuple<string, string, double>>& routes,
        string origin,
        double k
    ) {
        unordered_map<string, vector<pair<string, double>>> graph;

        // Build undirected graph
        for (auto& r : routes) {
            string u, v;
            double time;
            tie(u, v, time) = r;
            graph[u].push_back({v, time});
            graph[v].push_back({u, time});
        }

        // Min heap: (currentTime, city)
        priority_queue<
            pair<double, string>,
            vector<pair<double, string>>,
            greater<pair<double, string>>
        > pq;

        unordered_map<string, double> dist;

        for (auto& p : graph) {
            dist[p.first] = DBL_MAX;
        }

        dist[origin] = 0;
        pq.push({0, origin});

        // Dijkstra’s Algorithm
        while (!pq.empty()) {
            auto [currTime, city] = pq.top();
            pq.pop();

            if (currTime > dist[city])
                continue;

            for (auto& [next, travelTime] : graph[city]) {
                double newTime = currTime + travelTime;

                // Add stopover cost (1 hour) for intermediate cities
                if (city != origin)
                    newTime += 1;

                if (newTime < dist[next]) {
                    dist[next] = newTime;
                    pq.push({newTime, next});
                }
            }
        }

        // Count destinations within k hours (excluding origin)
        int count = 0;
        for (auto& p : dist) {
            if (p.first != origin && p.second <= k)
                count++;
        }

        return count;
    }
};

int main() {
    Solution s;

    vector<tuple<string, string, double>> routes = {
        {"Boston", "New York", 4},
        {"New York", "Philadelphia", 2},
        {"Boston", "Newport", 1.5},
        {"Washington, D.C.", "Harper's Ferry", 1},
        {"Boston", "Portland", 2.5},
        {"Philadelphia", "Washington, D.C.", 2.5}
    };

    cout << s.vacationDestinations(routes, "New York", 5) << endl; // 2
    cout << s.vacationDestinations(routes, "New York", 7) << endl; // 4
    cout << s.vacationDestinations(routes, "New York", 8) << endl; // 6

    return 0;
}
